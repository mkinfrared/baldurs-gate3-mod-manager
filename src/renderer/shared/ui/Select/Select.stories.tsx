/* eslint-disable react-hooks/rules-of-hooks */
import { Meta, StoryObj } from "@storybook/react";
import { useState } from "react";

import Select from "./Select";
import css from "./Story.module.scss";

type Story = StoryObj<typeof meta>;

const meta = {
  title: "UI/Select",
  component: Select,
  tags: ["autodocs"],
} satisfies Meta<typeof Select>;

const Default: Story = {
  args: {
    getActiveOption: () => false,
    getOptionLabel: () => null,
    options: [],
  },
  render: (args) => {
    const [selectedOption, setSelectedOption] = useState<string>();

    const states = [
      "Alabama",
      "Alaska",
      "Arizona",
      "Arkansas",
      "California",
      "Colorado",
      "Connecticut",
      "Delaware",
      "Florida",
      "Georgia",
      "Hawaii",
      "Idaho",
      "Illinois",
      "Indiana",
      "Iowa",
      "Kansas",
      "Kentucky",
      "Louisiana",
      "Maine",
      "Maryland",
      "Massachusetts",
      "Michigan",
      "Minnesota",
      "Mississippi",
      "Missouri",
      "Montana",
      "Nebraska",
      "Nevada",
      "New Hampshire",
      "New Jersey",
      "New Mexico",
      "New York",
      "North Carolina",
      "North Dakota",
      "Ohio",
      "Oklahoma",
      "Oregon",
      "Pennsylvania",
      "Rhode Island",
      "South Carolina",
      "South Dakota",
      "Tennessee",
      "Texas",
      "Utah",
      "Vermont",
      "Virginia",
      "Washington",
      "West Virginia",
      "Wisconsin",
      "Wyoming",
    ];

    const getOptionLabel = (value?: string) => value ?? "";
    const getActiveOption = (value: string) => value === selectedOption;

    return (
      <div className={css.Story}>
        <Select
          {...args}
          options={states}
          getOptionLabel={getOptionLabel}
          getActiveOption={getActiveOption}
          onChange={setSelectedOption}
          value={selectedOption}
        />
      </div>
    );
  },
};

export { Default };

export default meta;
