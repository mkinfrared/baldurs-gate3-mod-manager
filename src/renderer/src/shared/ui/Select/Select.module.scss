@import "../../styles/input-base";
@import "../../styles/vars";

.Select {
  @extend %input-base;

  .container {
    svg {
      transition: transform 250ms ease-in-out;
    }

    &[aria-expanded="true"] {
      svg {
        transform: rotateX(180deg);
      }
    }
  }

  .container:focus-visible {
    @include focus-visible;
  }
}
