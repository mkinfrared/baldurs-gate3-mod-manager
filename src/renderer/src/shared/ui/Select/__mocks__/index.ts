export { default } from "./Select.mock";
