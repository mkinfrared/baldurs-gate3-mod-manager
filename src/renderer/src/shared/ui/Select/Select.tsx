import { uniqueId } from "lodash-es";
import {
  ChangeEventHandler,
  KeyboardEventHandler,
  useMemo,
  useRef,
  useState,
} from "react";

import ArrowLeft from "@renderer/shared/assets/icons/arrow-down.svg";
import {
  classNames,
  getNodeText,
  isHtmlElement,
} from "@renderer/shared/lib/helpers";
import Heading from "@renderer/shared/ui/Heading";
import Menu from "@renderer/shared/ui/Menu";
import MenuItem from "@renderer/shared/ui/MenuItem";
import Spinner from "@renderer/shared/ui/Spinner";
import Text from "@renderer/shared/ui/Text";

import css from "./Select.module.scss";
import { SelectProps } from "./Select.type";

const Select = <T,>({
  className,
  disabled,
  error,
  filterable,
  getOptionLabel,
  getActiveOption,
  isLoading,
  label,
  name,
  options,
  onChange,
  placeholder,
  prefix,
  value,
}: SelectProps<T>) => {
  const [searchValue, setSearchValue] = useState<string>();
  const [isListboxOpen, setIsListboxOpen] = useState(false);
  const listboxId = useMemo(() => uniqueId("listbox"), []);
  const containerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  const closeListbox = () => {
    setIsListboxOpen(false);
  };

  const handleEnterDown = () => {
    if (isListboxOpen) {
      containerRef.current?.focus();
    }
  };

  const handleArrowPress = (key: "ArrowDown" | "ArrowUp") => {
    if (key === "ArrowDown") {
      setIsListboxOpen(true);
    }

    const listbox = document.getElementById(listboxId);

    if (!listbox) {
      return;
    }

    const element = listbox.querySelector('[tabindex="0"]');
    const { firstChild } = listbox;

    if (isHtmlElement(element)) {
      element.focus();

      return;
    }

    // If no element has tabIndex equal to 0
    // make the first element interactive
    if (isHtmlElement(firstChild)) {
      firstChild.tabIndex = 0;

      firstChild.focus();
    }
  };

  const handleKeyDown: KeyboardEventHandler = (event) => {
    const { key } = event;

    if (key === "Enter") {
      handleEnterDown();
    }

    if (key === "ArrowDown") {
      handleArrowPress("ArrowDown");
    }

    if (key === "ArrowUp") {
      handleArrowPress("ArrowUp");
    }

    if (key === "ArrowLeft" || key === "ArrowRight") {
      inputRef.current?.focus();
    }

    if (key === "Escape") {
      setIsListboxOpen(false);

      containerRef.current?.focus();
    }
  };

  const handleMenuItemClick = (itemValue?: T) => {
    onChange(itemValue);

    setSearchValue(undefined);

    closeListbox();
  };

  const handleInputChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    setSearchValue(event.target.value);
  };

  const getSuffix = () => {
    if (isLoading) {
      return <Spinner />;
    }

    return <ArrowLeft />;
  };

  const menuItems = options
    .filter((option) => {
      if (!searchValue) {
        return true;
      }

      const regex = new RegExp(searchValue, "i");
      const optionLabel = getOptionLabel(option);
      const optionText = getNodeText(optionLabel);

      return regex.test(optionText.toString());
    })
    .map((option, index) => (
      <MenuItem
        key={(option as any).id ?? index}
        onClick={handleMenuItemClick}
        isSelected={getActiveOption(option)}
        value={option}
      >
        {getOptionLabel(option)}
      </MenuItem>
    ));

  return (
    <label className={classNames(css.Select, className)} data-testid="Select">
      {label && (
        <Heading variant="h4" className={css.label}>
          {label}
        </Heading>
      )}
      <div
        aria-controls={listboxId}
        aria-expanded={isListboxOpen}
        className={css.container}
        ref={containerRef}
        onKeyDown={handleKeyDown}
        role="combobox"
        tabIndex={0}
      >
        {prefix && <div className={css.prefix}>{prefix}</div>}
        <input
          autoComplete="chrome-off"
          disabled={disabled}
          name={name}
          onChange={handleInputChange}
          onFocus={() => setIsListboxOpen(true)}
          placeholder={placeholder}
          readOnly={filterable}
          ref={inputRef}
          type="text"
          value={searchValue ?? getNodeText(getOptionLabel(value))}
        />
        <div className={css.suffix}>{getSuffix()}</div>
        {isListboxOpen && (
          <Menu
            id={listboxId}
            comboboxRef={containerRef}
            onClose={closeListbox}
          >
            {menuItems}
          </Menu>
        )}
      </div>
      <div className={css.errorContainer}>
        <Text className={css.error} size="small">
          {error}
        </Text>
      </div>
    </label>
  );
};

export default Select;
