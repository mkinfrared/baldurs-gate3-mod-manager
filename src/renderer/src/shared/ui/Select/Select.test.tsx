import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import Select from "./Select";

describe("<Select />", () => {
  const options = [42];
  const getActiveOption = () => false;
  const getOptionLabel = () => "marklar";
  const onChange = vi.fn();

  const Component = (
    <Select
      getActiveOption={getActiveOption}
      getOptionLabel={getOptionLabel}
      onChange={onChange}
      options={options}
    />
  );

  it("should be defined", () => {
    expect(Select).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Select");

    expect(element).toBeDefined();
  });
});
