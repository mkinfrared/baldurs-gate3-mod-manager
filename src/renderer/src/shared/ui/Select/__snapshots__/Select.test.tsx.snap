// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Select /> > should match the snapshot 1`] = `
<div>
  <label
    class="Select"
    data-testid="Select"
  >
    <div
      aria-controls="listbox1"
      aria-expanded="false"
      class="container"
      role="combobox"
      tabindex="0"
    >
      <input
        autocomplete="chrome-off"
        type="text"
        value="marklar"
      />
      <div
        class="suffix"
      >
        <svg>
          arrow-down.svg
        </svg>
      </div>
    </div>
    <div
      class="errorContainer"
    >
      <p
        class="Text error small"
        data-testid="Text"
      />
    </div>
  </label>
</div>
`;
